### PR TITLE
Pathfinder based cost function

### PR DESCRIPTION
**Pull request by ombhilare999**
Please find the changes done in this PR as follows:

 - The `width` attribute in the application graph has been renamed to `h_width` because `width` is a reserved keyword in Graphviz. Using `width` caused the cells to stretch, making the visualization of the `.dot` file more challenging.

- **Change of Cost Function**:
  - Previously, the cost function implemented in UGRAMM used a partial pathfinder approach. 
  - In this PR, I have incorporated a routability-driven pathfinder approach, similar to the one used in VPR.
  - The updated cost function is detailed below and can be observed in the `calculate_cost()` function in `routing.cpp`.
  - **TODO:** Implement BendCost will be implemented in the next iteration. 
image

- **Multi-fanout routing bug**:
  - In scenarios where sinks of a node are randomly allocated to the same slot, they tend to remain in that slot because it results in the best and similar cost for both nodes.  
  - This issue can be addressed by allocating a random node only if the node itself has not been utilized previously.

### CODE REVIEW

**Comment by ombhilare999** [comment]
Hi @HamasWaqar,
Can we merge this now so that I can update the exercise according to this?

File: src/UGRAMM.cpp [hunk 1]
```diff
@@ -14,11 +14,12 @@
 //-------------------------------------------------------------------//
 
 // Pathefinder cost parameters:
-float PFac = 1; // Congestion cost factor
-float HFac = 1; // History cost factor
+int iterCount = 0;
+float PFac = 1;       // Congestion cost factor
+float HFac = 1;       // History cost factor
 float pfac_mul = 1.1; //Multiplier for present congestion cost [defaults to 1.1]
 float hfac_mul = 2;   //Multiplier for history congestion cost [defaults to 2]
```
Comment (HamasWaqar) [inline src/UGRAMM.cpp]: should we make a comment in our documentation that hfac_mul currently is just hfac? Basically, what we discussed in the meeting earlier

**Comment by HamasWaqar** [comment]
Other than above, all looks good

File: src/UGRAMM.cpp [hunk 2]
```diff
@@ -508,7 +510,7 @@ int main(int argc, char **argv)
       ("drc_disable", po::bool_switch(&drc_disable), "disable DRC [optional]")
       ("max_iter", po::value<int>(&max_iter), "Maximum number of iterations UGRAMM will run [optional; defaults to 40]" )
       ("pfac_mul", po::value<float>(&pfac_mul), "Multiplier for present congestion cost [optional; defaults to 1.1]")
-      ("hfac_mul", po::value<float>(&hfac_mul), "Multiplier for history congestion cost [optional; defaults to 2]")
+      ("hfac_mul", po::value<float>(&hfac_mul), "Multiplier for history congestion cost [optional; defaults to 1]")
```
[the remark this answers is not in the document]
Comment (ombhilare999) [reply src/UGRAMM.cpp:513]: Sounds good. Let's create a page on this on the Docs website.

File: src/UGRAMM.cpp (hunk 1 above)
Comment (ombhilare999) [reply src/UGRAMM.cpp]: Sounds good. Let's create a page on this on the Docs website.

**Pull request merged**